src/Lean/Elab/PreDefinition/Structural.lean

```diff
@@ -392,6 +449,41 @@ private def mkBRecOn (recFnName : Name) (recArgInfo : RecArgInfo) (value : Expr)
       let Farg         ← mkLambdaFVars (recArgInfo.indIndices ++ #[major, below] ++ otherArgs) valueNew
       let brecOn       := mkApp brecOn Farg
       pure $ mkAppN brecOn otherArgs
+      
+private def mkIndPredBRecOn (recFnName : Name) (recArgInfo : RecArgInfo) (value : Expr) : M Expr := do
+  let type  := (← inferType value).headBeta
+  let major := recArgInfo.ys[recArgInfo.pos]
+  let otherArgs := recArgInfo.ys.filter fun y => y != major && !recArgInfo.indIndices.contains y
+  trace[Elab.definition.structural] "fixedParams: {recArgInfo.fixedParams}, otherArgs: {otherArgs}"
+  let motive ← mkForallFVars otherArgs type
+  let motive ← mkLambdaFVars (recArgInfo.indIndices.push major) motive
+  trace[Elab.definition.structural] "brecOn motive: {motive}"
+  let brecOn :=
+      Lean.mkConst (mkBRecOnName recArgInfo.indName) recArgInfo.indLevels
+  let brecOn := mkAppN brecOn recArgInfo.indParams
+  let brecOn := mkApp brecOn motive
+  let brecOn := mkAppN brecOn recArgInfo.indIndices
+  let brecOn := mkApp brecOn major
+  check brecOn
+  let brecOnType ← inferType brecOn
+  trace[Elab.definition.structural] "brecOn     {brecOn}"
+  trace[Elab.definition.structural] "brecOnType {brecOnType}"
+  forallBoundedTelescope brecOnType (some 1) fun F _ => do
+    let F := F[0]
+    let FType ← inferType F
+    trace[Elab.definition.structural] "FType: {FType}"
+    let FType ← instantiateForall FType recArgInfo.indIndices
+    let FType ← instantiateForall FType #[major]
+    forallBoundedTelescope FType (some 1) fun below _ => do
+      let lctx ← getLCtx
+      let lctx := lctx.erase F.fvarId!
+      withTheReader Meta.Context (fun ctx => { ctx with lctx }) do
```

Comment:
We can avoid this hack by restricting the scope of `forallBoundedTelescope brecOnType (some 1) fun F _ => do`. 
As far as I can tell, you are using it to extract `FType`, correct?
If this is the case, we can write
```lean
let FType ←forallBoundedTelescope brecOnType (some 1) fun F _ => do
  let F := F[0]
  let FType ← inferType F
  trace[Elab.definition.structural] "FType: {FType}"
  let FType ← instantiateForall FType recArgInfo.indIndices
  instantiateForall FType #[major]
-- The variable `F` is not in the local context anymore.
forallBoundedTelescope FType (some 1) fun below _ => do
  let main ← mkFreshExprSyntheticOpaqueMVar FType
  ...
```

Reply:
I'll double-check that. The root cause was, that this copied code puts an `ih : FType` into the local context and later, when I use the local context to build the recursive call, it uses this `ih`. But that `ih` doesn't exist in the actual `brecOn` call. That's why it must go. I think this would work fine with your approach.

Reply:
Could you please add a comment with this explanation?
I think this idiom is not used anywhere in the codebase.